packages/jupyter-ai/src/chat_handler.ts

```diff
@@ -106,23 +96,84 @@ export class ChatHandler implements IDisposable {
     }
   }
 
-  private _onMessage(message: AiService.Message): void {
+  get history(): AiService.ChatHistory {
+    return {
+      messages: this._messages,
+      pending_messages: this._pendingMessages
+    };
+  }
+
+  get historyChanged(): Signal<this, AiService.ChatHistory> {
+    return this._historyChanged;
+  }
+
+  private _onMessage(newMessage: AiService.Message): void {
     // resolve promise from `sendMessage()`
-    if (message.type === 'human' && message.client.id === this.id) {
-      this._sendResolverQueue.shift()?.(message.id);
+    if (newMessage.type === 'human' && newMessage.client.id === this.id) {
+      this._sendResolverQueue.shift()?.(newMessage.id);
     }
 
     // resolve promise from `replyFor()` if it exists
     if (
-      message.type === 'agent' &&
-      message.reply_to in this._replyForResolverDict
+      newMessage.type === 'agent' &&
+      newMessage.reply_to in this._replyForResolverDict
     ) {
-      this._replyForResolverDict[message.reply_to](message);
-      delete this._replyForResolverDict[message.reply_to];
+      this._replyForResolverDict[newMessage.reply_to](newMessage);
+      delete this._replyForResolverDict[newMessage.reply_to];
     }
 
     // call listeners in serial
-    this._listeners.forEach(listener => listener(message));
+    this._listeners.forEach(listener => listener(newMessage));
+
+    // append message to chat history. this block should always set `_messages`
+    // or `_pendingMessages` to a new array instance rather than modifying
+    // in-place so consumer React components re-render.
+    switch (newMessage.type) {
+      case 'connection':
+        break;
+      case 'clear':
+        this._messages = [];
+        break;
+      case 'pending':
+        this._pendingMessages = [...this._pendingMessages, newMessage];
+        break;
+      case 'close-pending':
+        this._pendingMessages = this._pendingMessages.filter(
+          p => p.id !== newMessage.id
+        );
+        break;
+      case 'agent-stream-chunk': {
+        const target = newMessage.id;
+        const streamMessage = this._messages.find<AiService.AgentStreamMessage>(
+          (m): m is AiService.AgentStreamMessage =>
+            m.type === 'agent-stream' && m.id === target
+        );
+        if (!streamMessage) {
+          console.error(
+            `Received stream chunk with ID ${target}, but no agent-stream message with that ID exists. ` +
+              'Ignoring this stream chunk.'
+          );
+          break;
+        }
+
+        streamMessage.body += newMessage.content;
+        if (newMessage.stream_complete) {
+          streamMessage.complete = true;
+        }
+        this._messages = [...this._messages];
```

Comment:
Why is this required?

Reply:
This is done so consumer React components re-render when directly calling `setMessages()` on the returned `history.messages` object that is emitted in the `historyChanged` signal. I added a comment at the top of the `switch` block for context:


```ts
    // append message to chat history. this block should always set `_messages`
    // or `_pendingMessages` to a new array instance rather than modifying
    // in-place so consumer React components re-render.
    switch (newMessage.type) {
       ...
    }
```